File: internal/updater/updater.go
package updater

import (
	"archive/zip"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/pyrolis-app/pyrolis-connector/internal/config"
)

const (
	githubReleasesAPI = "https://api.github.com/repos/pyrolis-app/pyrolis-connector/releases/latest"
	checkInterval     = 6 * time.Hour
)

// Status values
const (
	StatusIdle        = "idle"
	StatusAvailable   = "available"
	StatusDownloading = "downloading"
	StatusReady       = "ready"
	StatusApplying    = "applying"
	StatusError       = "error"
)

// State represents the current update state.
type State struct {
	Status           string    `json:"status"`
	AvailableVersion string    `json:"available_version,omitempty"`
	DownloadURL      string    `json:"download_url,omitempty"`
	Checksum         string    `json:"checksum,omitempty"`
	DownloadPath     string    `json:"download_path,omitempty"`
	Error            string    `json:"error,omitempty"`
	CheckedAt        time.Time `json:"checked_at,omitempty"`
	CurrentVersion   string    `json:"current_version"`
}

// Updater manages self-updates.
type Updater struct {
	mu      sync.RWMutex
	state   State
	version string
	stopCh  chan struct{}
}

// New creates a new Updater.
func New(version string) *Updater {
	return &Updater{
		version: version,
		state:   State{Status: StatusIdle, CurrentVersion: version},
		stopCh:  make(chan struct{}),
	}
}

// Start begins the periodic update check loop.
func (u *Updater) Start() {
	// Initial check after a short delay
	go func() {
		time.Sleep(30 * time.Second)
		u.CheckGitHub()
	}()

	go func() {
		ticker := time.NewTicker(checkInterval)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				u.mu.RLock()
				s := u.state.Status
				u.mu.RUnlock()
				if s == StatusIdle {
					u.CheckGitHub()
				}
			case <-u.stopCh:
				return
			}
		}
	}()
}

// Stop halts periodic checks.
func (u *Updater) Stop() {
	close(u.stopCh)
}

// GetState returns the current update state.
func (u *Updater) GetState() State {
	u.mu.RLock()
	defer u.mu.RUnlock()
	s := u.state
	s.CurrentVersion = u.version
	return s
}

// NotifyAvailable is called when the cloud pushes an update notification.
func (u *Updater) NotifyAvailable(version, downloadURL, checksum string) {
	cfg := config.Get()
	if cfg != nil && !cfg.Settings.AllowRemoteUpdates {
		slog.Info("Ignoring remote update push (remote updates disabled)")
		return
	}

	if !newerVersion(version, u.version) {
		slog.Debug("Ignoring update, already at or newer", "available", version, "current", u.version)
		return
	}

	slog.Info("Update available", "version", version)

	u.mu.Lock()
	u.state = State{
		Status:           StatusAvailable,
		AvailableVersion: version,
		DownloadURL:      downloadURL,
		Checksum:         checksum,
		CurrentVersion:   u.version,
	}
	u.mu.Unlock()

	// Auto-act based on mode
	if cfg != nil {
		switch cfg.Settings.AutoApplyMode {
		case "auto":
			slog.Info("Auto-install: downloading and applying")
			go u.DownloadAndApply()
		case "download":
			slog.Info("Auto-download: downloading")
			go u.Download()
		default:
			slog.Info("Manual mode: waiting for user action")
		}
	}
}

// CheckGitHub checks GitHub releases for a newer version.
func (u *Updater) CheckGitHub() {
	slog.Debug("Checking GitHub for updates")

	u.mu.Lock()
	u.state.CheckedAt = time.Now()
	u.mu.Unlock()

	req, _ := http.NewRequest("GET", githubReleasesAPI, nil)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "pyrolis-connector")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		slog.Warn("GitHub update check failed", "error", err)
		u.mu.Lock()
		u.state.Error = fmt.Sprintf("Check failed: %v", err)
		u.mu.Unlock()
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		slog.Warn("GitHub API returned", "status", resp.StatusCode)
		return
	}

	var release struct {
		TagName string `json:"tag_name"`
		Assets  []struct {
			Name               string `json:"name"`
			BrowserDownloadURL string `json:"browser_download_url"`
		} `json:"assets"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		slog.Warn("Failed to parse GitHub release", "error", err)
		return
	}

	version := strings.TrimPrefix(release.TagName, "pyrolis-connector-v")
	target := PlatformTarget()

	var downloadURL, checksumURL string
	for _, a := range release.Assets {
		if strings.Contains(a.Name, target) && !strings.HasSuffix(a.Name, ".txt") {
			downloadURL = a.BrowserDownloadURL
		}
		if a.Name == "SHA256SUMS.txt" {
			checksumURL = a.BrowserDownloadURL
		}
	}

	if downloadURL == "" {
		slog.Debug("No binary found for platform", "target", target)
		return
	}

	checksum := fetchChecksum(checksumURL, target)

	if newerVersion(version, u.version) {
		slog.Info("GitHub check: update available", "version", version, "current", u.version)
		u.mu.Lock()
		u.state = State{
			Status:           StatusAvailable,
			AvailableVersion: version,
			DownloadURL:      downloadURL,
			Checksum:         checksum,
			CheckedAt:        time.Now(),
			CurrentVersion:   u.version,
		}
		u.mu.Unlock()
	} else {
		slog.Debug("Already up to date", "current", u.version)
		u.mu.Lock()
		u.state.Status = StatusIdle
		u.state.Error = ""
		u.mu.Unlock()
	}
}

// Download downloads the available update.
func (u *Updater) Download() {
	u.mu.RLock()
	if u.state.Status != StatusAvailable || u.state.DownloadURL == "" {
		u.mu.RUnlock()
		return
	}
	url := u.state.DownloadURL
	u.mu.RUnlock()

	u.mu.Lock()
	u.state.Status = StatusDownloading
	u.state.Error = ""
	u.mu.Unlock()

	u.mu.RLock()
	checksum := u.state.Checksum
	u.mu.RUnlock()

	slog.Info("Downloading update", "url", url)

	path, err := downloadFile(url, checksum)
	if err != nil {
		slog.Error("Download failed", "error", err)
		u.mu.Lock()
		u.state.Status = StatusError
		u.state.Error = fmt.Sprintf("Download failed: %v", err)
		u.mu.Unlock()
		return
	}

	u.mu.Lock()
	u.state.Status = StatusReady
	u.state.DownloadPath = path
	u.mu.Unlock()

	slog.Info("Download complete", "path", path)
}

// DownloadAndApply downloads and then applies the update.
func (u *Updater) DownloadAndApply() {
	u.Download()

	u.mu.RLock()
	if u.state.Status != StatusReady {
		u.mu.RUnlock()
		return
	}
	u.mu.RUnlock()

	u.Apply()
}

// Apply applies a downloaded update.
func (u *Updater) Apply() {
	u.mu.Lock()
	if u.state.Status != StatusReady || u.state.DownloadPath == "" {
		u.mu.Unlock()
		return
	}
	path := u.state.DownloadPath
	u.state.Status = StatusApplying
	u.mu.Unlock()

	slog.Info("Applying update", "path", path)

	// Platform-specific binary replacement + restart
	if err := selfUpdate(path); err != nil {
		slog.Error("Self-update failed", "error", err)
		os.Remove(path)
		u.mu.Lock()
		u.state.Status = StatusError
		u.state.Error = err.Error()
		u.mu.Unlock()
	}
	// If selfUpdate succeeds, the process is replaced and we never get here
}

// Dismiss resets the update state.
func (u *Updater) Dismiss() {
	u.mu.Lock()
	defer u.mu.Unlock()
	if u.state.DownloadPath != "" {
		os.Remove(u.state.DownloadPath)
	}
	u.state = State{Status: StatusIdle, CheckedAt: u.state.CheckedAt, CurrentVersion: u.version}
}

// PlatformTarget returns the platform identifier for asset matching.
func PlatformTarget() string {
	os := runtime.GOOS
	arch := runtime.GOARCH
	return fmt.Sprintf("%s-%s", os, arch)
}

func downloadFile(url, checksum string) (string, error) {
	// Download to a temp file
	tmpFile, err := os.CreateTemp("", "pyrolis-connector-update-*")
	if err != nil {
		return "", err
	}
	defer tmpFile.Close()

	resp, err := http.Get(url)
	if err != nil {
		os.Remove(tmpFile.Name())
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		os.Remove(tmpFile.Name())
		return "", fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	if _, err := io.Copy(tmpFile, resp.Body); err != nil {
		os.Remove(tmpFile.Name())
		return "", err
	}
	tmpFile.Close()

	// Verify checksum on the downloaded archive before extracting
	if checksum != "" {
		if err := verifyChecksum(tmpFile.Name(), checksum); err != nil {
			os.Remove(tmpFile.Name())
			return "", fmt.Errorf("checksum: %w", err)
		}
	}

	// If the download is a zip, extract the binary from it
	if strings.HasSuffix(url, ".zip") {
		extracted, err := extractBinaryFromZip(tmpFile.Name())
		os.Remove(tmpFile.Name())
		if err != nil {
			return "", fmt.Errorf("extract zip: %w", err)
		}
		return extracted, nil
	}

	return tmpFile.Name(), nil
}

// extractBinaryFromZip opens a zip archive and extracts the pyrolis-connector binary.
func extractBinaryFromZip(zipPath string) (string, error) {
	r, err := zip.OpenReader(zipPath)
	if err != nil {
		return "", err
	}
	defer r.Close()

	binaryName := "pyrolis-connector"
	if runtime.GOOS == "windows" {
		binaryName = "pyrolis-connector.exe"
	}

	for _, f := range r.File {
		name := filepath.Base(f.Name)
		if name != binaryName {
			continue
		}

		rc, err := f.Open()
		if err != nil {
			return "", err
		}
		defer rc.Close()

		tmpFile, err := os.CreateTemp("", "pyrolis-connector-bin-*")
		if err != nil {
			return "", err
		}

		if _, err := io.Copy(tmpFile, rc); err != nil {
			tmpFile.Close()
			os.Remove(tmpFile.Name())
			return "", err
		}
		tmpFile.Close()
		return tmpFile.Name(), nil
	}

	return "", fmt.Errorf("binary %q not found in zip", binaryName)
}

func fetchChecksum(checksumURL, target string) string {
	if checksumURL == "" {
		return ""
	}
	resp, err := http.Get(checksumURL)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	for _, line := range strings.Split(string(body), "\n") {
		parts := strings.Fields(line)
		if len(parts) == 2 && strings.Contains(parts[1], target) {
			return "sha256:" + parts[0]
		}
	}
	return ""
}

func verifyChecksum(path, expected string) error {
	if !strings.HasPrefix(expected, "sha256:") {
		return nil
	}
	expectedHex := strings.TrimPrefix(expected, "sha256:")

	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return err
	}

	actual := hex.EncodeToString(h.Sum(nil))
	if !strings.EqualFold(actual, expectedHex) {
		return fmt.Errorf("checksum mismatch: expected %s, got %s", expectedHex, actual)
	}
	return nil
}

func newerVersion(available, current string) bool {
	// Simple semver comparison: split on "." and compare numerically
	av := parseVersion(available)
	cv := parseVersion(current)
	if av == nil || cv == nil {
		return false
	}
	for i := 0; i < 3; i++ {
		if av[i] > cv[i] {
			return true
		}
		if av[i] < cv[i] {
			return false
		}
	}
	return false
}

func parseVersion(v string) []int {
	// Strip git describe suffix like "-4-gabcdef"
	if idx := strings.Index(v, "-"); idx > 0 {
		rest := v[idx+1:]
		if len(rest) > 0 && rest[0] >= '0' && rest[0] <= '9' {
			v = v[:idx]
		}
	}
	v = strings.TrimSpace(v)

	parts := strings.Split(v, ".")
	if len(parts) < 3 {
		return nil
	}
	nums := make([]int, 3)
	for i := 0; i < 3; i++ {
		fmt.Sscanf(parts[i], "%d", &nums[i])
	}
	return nums
}

File: lib/pyrolis_connector/updater.ex
defmodule PyrolisConnector.Updater do
  @moduledoc """
  Self-update manager for the Pyrolis Connector.

  Downloads new binary releases from GitHub, verifies their SHA-256 checksum,
  replaces the running binary, and restarts the application.

  ## Update flow

  1. Cloud pushes `"update_available"` via WebSocket (or user clicks "Check for updates")
  2. Updater downloads the new binary to a temp file
  3. Verifies the SHA-256 checksum
  4. Replaces the current binary (backup kept as `<binary>.bak`)
  5. Restarts the application via `System.restart/0`

  ## State

  The updater tracks its current status as one of:
  - `:idle` — No update activity
  - `:available` — Update available, not yet downloading
  - `:downloading` — Download in progress
  - `:ready` — Downloaded and verified, ready to apply
  - `:applying` — Replacing binary and restarting
  - `:error` — Something went wrong (see `error` field)
  """

  use GenServer

  require Logger

  @github_releases_url "https://github.com/pyrolis-app/pyrolis-connector/releases"
  @check_interval_ms :timer.hours(6)

  defstruct [
    :status,
    :available_version,
    :download_url,
    :checksum,
    :download_path,
    :error,
    :checked_at
  ]

  # Client API

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @doc "Returns the current update status."
  def status do
    case Process.whereis(__MODULE__) do
      nil -> %__MODULE__{status: :idle}
      pid ->
        try do
          GenServer.call(pid, :status, 5_000)
        catch
          :exit, _ -> %__MODULE__{status: :idle}
        end
    end
  end

  @doc "Notify the updater that a new version is available (from cloud push)."
  def notify_available(version, download_url, checksum) do
    case Process.whereis(__MODULE__) do
      nil -> :ok
      pid -> GenServer.cast(pid, {:update_available, version, download_url, checksum})
    end
  end

  @doc "Check GitHub releases for a newer version."
  def check_now do
    case Process.whereis(__MODULE__) do
      nil -> {:error, :not_running}
      pid -> GenServer.cast(pid, :check_github)
    end
  end

  @doc "Download the available update."
  def download do
    case Process.whereis(__MODULE__) do
      nil -> {:error, :not_running}
      pid -> GenServer.cast(pid, :download)
    end
  end

  @doc "Apply a downloaded update (replace binary + restart)."
  def apply_update do
    case Process.whereis(__MODULE__) do
      nil -> {:error, :not_running}
      pid -> GenServer.cast(pid, :apply_update)
    end
  end

  @doc "Check if remote (cloud-pushed) updates are allowed."
  def remote_updates_allowed? do
    PyrolisConnector.State.get_setting("allow_remote_updates") != "false"
  end

  @doc "Enable or disable remote update pushes."
  def set_remote_updates(enabled) when is_boolean(enabled) do
    PyrolisConnector.State.save_setting("allow_remote_updates", to_string(enabled))
  end

  @doc "Dismiss the current update notification."
  def dismiss do
    case Process.whereis(__MODULE__) do
      nil -> :ok
      pid -> GenServer.cast(pid, :dismiss)
    end
  end

  # Server

  @impl true
  def init(_opts) do
    # Schedule periodic check
    Process.send_after(self(), :periodic_check, @check_interval_ms)

    {:ok, %__MODULE__{status: :idle}}
  end

  @impl true
  def handle_call(:status, _from, state) do
    {:reply, state, state}
  end

  @impl true
  def handle_cast({:update_available, version, download_url, checksum}, state) do
    current = PyrolisConnector.version()

    if newer_version?(version, current) do
      Logger.info("Update available: v#{version} (current: v#{current})")

      {:noreply,
       %{state | status: :available, available_version: version, download_url: download_url, checksum: checksum, error: nil}}
    else
      Logger.debug("Ignoring update v#{version}, already at v#{current}")
      {:noreply, state}
    end
  end

  def handle_cast(:check_github, state) do
    me = self()

    Task.start(fn ->
      result = check_github_releases()
      send(me, {:github_check_result, result})
    end)

    {:noreply, %{state | checked_at: DateTime.utc_now()}}
  end

  def handle_cast(:download, %{status: :available, download_url: url} = state) when is_binary(url) do
    me = self()
    Logger.info("Starting download from #{url}")

    Task.start(fn ->
      result = download_binary(url)
      send(me, {:download_result, result})
    end)

    {:noreply, %{state | status: :downloading, error: nil}}
  end

  def handle_cast(:download, state) do
    {:noreply, state}
  end

  def handle_cast(:apply_update, %{status: :ready, download_path: path} = state) when is_binary(path) do
    Logger.info("Applying update from #{path}")

    case apply_binary_update(path, state.checksum) do
      :ok ->
        Logger.info("Update applied successfully, restarting...")
        {:noreply, %{state | status: :applying}}

      {:error, reason} ->
        Logger.error("Failed to apply update: #{inspect(reason)}")
        {:noreply, %{state | status: :error, error: to_string(reason)}}
    end
  end

  def handle_cast(:apply_update, state) do
    {:noreply, state}
  end

  def handle_cast(:dismiss, state) do
    # Clean up any downloaded file
    if state.download_path, do: File.rm(state.download_path)

    {:noreply, %__MODULE__{status: :idle, checked_at: state.checked_at}}
  end

  @impl true
  def handle_info({:github_check_result, {:ok, version, download_url, checksum}}, state) do
    current = PyrolisConnector.version()

    if newer_version?(version, current) do
      Logger.info("GitHub check: update available v#{version} (current: v#{current})")

      {:noreply,
       %{state | status: :available, available_version: version, download_url: download_url, checksum: checksum, error: nil}}
    else
      Logger.debug("GitHub check: already up to date (v#{current})")
      {:noreply, %{state | status: :idle, error: nil}}
    end
  end

  def handle_info({:github_check_result, {:error, reason}}, state) do
    Logger.warning("GitHub update check failed: #{inspect(reason)}")
    {:noreply, %{state | error: "Check failed: #{reason}"}}
  end

  def handle_info({:download_result, {:ok, path}}, state) do
    Logger.info("Download complete: #{path}")
    {:noreply, %{state | status: :ready, download_path: path, error: nil}}
  end

  def handle_info({:download_result, {:error, reason}}, state) do
    Logger.error("Download failed: #{inspect(reason)}")
    {:noreply, %{state | status: :error, error: "Download failed: #{reason}"}}
  end

  def handle_info(:periodic_check, state) do
    # Only auto-check if idle (not mid-update)
    if state.status == :idle do
      GenServer.cast(self(), :check_github)
    end

    Process.send_after(self(), :periodic_check, @check_interval_ms)
    {:noreply, state}
  end

  def handle_info(_msg, state) do
    {:noreply, state}
  end

  # Private helpers

  defp check_github_releases do
    # Use GitHub API to get latest release
    api_url = String.replace(@github_releases_url, "github.com", "api.github.com/repos") <> "/latest"

    case Req.get(api_url, headers: [{"accept", "application/vnd.github+json"}, {"user-agent", "pyrolis-connector"}], receive_timeout: 15_000) do
      {:ok, %{status: 200, body: body}} ->
        parse_github_release(body)

      {:ok, %{status: status}} ->
        {:error, "GitHub API returned #{status}"}

      {:error, reason} ->
        {:error, inspect(reason)}
    end
  end

  defp parse_github_release(body) do
    tag = body["tag_name"] || ""
    version = String.replace(tag, ~r/^pyrolis-connector-v/, "")

    target = platform_target()
    assets = body["assets"] || []

    # Find the binary asset for our platform
    binary_asset =
      Enum.find(assets, fn a ->
        name = a["name"] || ""
        String.contains?(name, target) and not String.ends_with?(name, ".txt")
      end)

    # Find SHA256SUMS.txt
    checksums_asset = Enum.find(assets, fn a -> a["name"] == "SHA256SUMS.txt" end)

    case binary_asset do
      nil ->
        {:error, "No binary found for platform #{target}"}

      asset ->
        checksum = fetch_checksum(checksums_asset, asset["name"])
        {:ok, version, asset["browser_download_url"], checksum}
    end
  end

  defp fetch_checksum(nil, _filename), do: nil

  defp fetch_checksum(checksums_asset, filename) do
    case Req.get(checksums_asset["browser_download_url"], receive_timeout: 10_000) do
      {:ok, %{status: 200, body: body}} when is_binary(body) ->
        body
        |> String.split("\n")
        |> Enum.find_value(fn line ->
          case String.split(line, ~r/\s+/, parts: 2) do
            [hash, ^filename] -> "sha256:#{hash}"
            _ -> nil
          end
        end)

      _ ->
        nil
    end
  end

  defp download_binary(url) do
    tmp_dir = System.tmp_dir!()
    tmp_path = Path.join(tmp_dir, "pyrolis-connector-update-#{:erlang.unique_integer([:positive])}")

    case Req.get(url, into: File.stream!(tmp_path), receive_timeout: 300_000) do
      {:ok, %{status: 200}} ->
        {:ok, tmp_path}

      {:ok, %{status: status}} ->
        File.rm(tmp_path)
        {:error, "Download returned HTTP #{status}"}

      {:error, reason} ->
        File.rm(tmp_path)
        {:error, inspect(reason)}
    end
  end

  defp apply_binary_update(download_path, expected_checksum) do
    # Verify checksum if provided
    checksum_ok =
      if expected_checksum do
        verify_checksum(download_path, expected_checksum)
      else
        :ok
      end

    case checksum_ok do
      {:error, _} = err ->
        return_and_cleanup(err, download_path)

      :ok ->
        apply_verified_binary(download_path)
    end
  end

  defp apply_verified_binary(download_path) do
    case current_binary_path() do
      {:burrito, bin_path} ->
        # In-place replacement of the single Burrito binary
        backup_path = bin_path <> ".bak"

        with :ok <- File.rename(bin_path, backup_path),
             :ok <- File.rename(download_path, bin_path),
             :ok <- make_executable(bin_path) do
          schedule_restart()
          :ok
        else
          {:error, reason} ->
            if File.exists?(backup_path) and not File.exists?(bin_path) do
              File.rename(backup_path, bin_path)
            end

            {:error, reason}
        end

      {:release, root} ->
        # Standard release: place the new binary next to the release root
        # so the user can switch to the standalone binary
        dest = Path.join(Path.dirname(root), "pyrolis-connector")
        backup = dest <> ".bak"

        if File.exists?(dest), do: File.rename(dest, backup)

        with :ok <- File.rename(download_path, dest),
             :ok <- make_executable(dest) do
          Logger.info("New binary placed at #{dest}")
          schedule_restart()
          :ok
        else
          {:error, reason} ->
            if File.exists?(backup) and not File.exists?(dest) do
              File.rename(backup, dest)
            end

            {:error, reason}
        end

      :unknown ->
        # Dev mode (mix run): just place the binary in cwd
        dest = Path.join(File.cwd!(), "pyrolis-connector")
        backup = dest <> ".bak"

        if File.exists?(dest), do: File.rename(dest, backup)

        with :ok <- File.rename(download_path, dest),
             :ok <- make_executable(dest) do
          Logger.info("New binary downloaded to #{dest}")
          Logger.info("Restart with: ./pyrolis-connector")
          :ok
        else
          {:error, reason} ->
            if File.exists?(backup) and not File.exists?(dest) do
              File.rename(backup, dest)
            end

            {:error, reason}
        end
    end
  end

  defp schedule_restart do
    Task.start(fn ->
      Process.sleep(1_000)

      case System.get_env("__BURRITO_BIN_PATH") do
        path when is_binary(path) ->
          # Re-exec the binary as a fresh OS process to avoid BEAM :low_entropy crash
          Logger.info("Re-launching #{path}...")
          Port.open({:spawn_executable, path}, [:binary, :exit_status, args: []])
          Process.sleep(500)
          System.halt(0)

        _ ->
          Logger.info("Restarting application...")
          System.restart()
      end
    end)
  end

  defp return_and_cleanup({:error, _} = err, path) do
    File.rm(path)
    err
  end

  defp verify_checksum(path, "sha256:" <> expected_hex) do
    actual_hex =
      File.stream!(path, 65_536)
      |> Enum.reduce(:crypto.hash_init(:sha256), fn chunk, acc ->
        :crypto.hash_update(acc, chunk)
      end)
      |> :crypto.hash_final()
      |> Base.encode16(case: :lower)

    if actual_hex == String.downcase(expected_hex) do
      :ok
    else
      {:error, "Checksum mismatch: expected #{expected_hex}, got #{actual_hex}"}
    end
  end

  defp verify_checksum(_path, _other), do: :ok

  defp current_binary_path do
    case System.get_env("__BURRITO_BIN_PATH") do
      path when is_binary(path) -> {:burrito, path}
      nil ->
        case System.get_env("RELEASE_ROOT") do
          root when is_binary(root) -> {:release, root}
          nil -> :unknown
        end
    end
  end

  defp make_executable(path) do
    case :os.type() do
      {:unix, _} -> File.chmod(path, 0o755)
      _ -> :ok
    end
  end

  defp platform_target do
    case :os.type() do
      {:win32, _} -> "windows"
      {:unix, :darwin} -> "darwin"
      {:unix, _} -> "linux"
    end
  end

  @doc false
  def newer_version?(available, current) do
    case {parse_version(available), parse_version(current)} do
      {{:ok, av}, {:ok, cv}} -> Version.compare(av, cv) == :gt
      _ -> false
    end
  end

  defp parse_version(v) do
    # Handle versions like "0.3.0-4-gabcdef" from git describe
    clean =
      v
      |> String.trim()
      |> String.replace(~r/-\d+-g[0-9a-f]+$/, "")

    Version.parse(clean)
  end
end
